### PR TITLE
More efficient handshake encoding

### PR DESCRIPTION
**Pull request by ctz**
This is showing a 2-3% improvement on resumption benchmarks for me.

### CODE REVIEW

**Review by djc — COMMENTED** [review]
Awesome! How did you discover this?

I would suggest changing the `NestedLength` type to hold on to the `bytes` reference (maybe `buf`?), and renaming it accordingly (maybe `ListBuffer`?). Could expose the buffer via `pub` field or a `get_mut()` getter, and add a `Drop` impl that does the `finish()` work.

**Comment by ctz** [comment]
> Awesome! How did you discover this?

I'm experimenting with improving how extensions are represented internally. Especially, to remove the need for a `HashMap` and manual checking to detect duplicate extensions.

> I would suggest changing the `NestedLength` type to hold on to the `bytes` reference (maybe `buf`?), and renaming it accordingly (maybe `ListBuffer`?). Could expose the buffer via `pub` field or a `get_mut()` getter, and add a `Drop` impl that does the `finish()` work.

Yes, this is a lot better. I was a bit nervous of using Drop, probably without good reason.

File: rustls/src/msgs/codec.rs [hunk 1]
```diff
@@ -270,9 +270,9 @@ impl<'a> LengthPrefixedBuffer<'a> {
     pub(crate) fn new(size_len: ListLength, buf: &'a mut Vec<u8>) -> LengthPrefixedBuffer<'a> {
         let len_offset = buf.len();
         buf.extend(match size_len {
-            ListLength::U8 => &[0][..],
-            ListLength::U16 => &[0, 0],
-            ListLength::U24 { .. } => &[0, 0, 0],
+            ListLength::U8 => &[0xff][..],
```
Comment (djc) [inline rustls/src/msgs/codec.rs:273]: Commit message has some interesting spellings of "interpreted" like "intepretted" and "interpretted".

**Review by djc — APPROVED** [review]
Cool!

File: rustls/src/msgs/codec.rs (hunk 1 above)
Comment (ctz) [reply rustls/src/msgs/codec.rs:273]: shameful 🙈

**Review by cpu — APPROVED** [review]
Cool!

**Pull request merged**